contracts/ligo/src/swaps/fa2_swap.mligo

```diff
@@ -0,0 +1,158 @@
+#import "../../fa2/fa2_interface.mligo" "FA2"
+
+(* ==== Types ==== *)
+
+type swap_id = nat
+
+type fa2_token =
+  [@layout:comb]
+  { token_id : FA2.token_id
+  ; amount : nat
+  }
+
+type fa2_assets =
+  [@layout:comb]
+  { fa2_address : address
+  ; tokens : fa2_token list
+  }
+
+type swap_offer =
+  [@layout:comb]
+  { assets_offered : fa2_assets list
+  ; assets_requested : fa2_assets list
+  }
+
+type swap_info =
+  [@layout:comb]
+  { swap_offer : swap_offer
+  ; seller : address
+  }
+
+type swap_entrypoints =
+  | Start of swap_offer
+  | Cancel of swap_id
+  | Accept of swap_id
+
+type swap_storage =
+  { next_swap_id : swap_id
+  ; swaps : (swap_id, swap_info) big_map
+  }
+
+type return = operation list * swap_storage
+
+(* ==== Values ==== *)
+
+let init_storage : swap_storage =
+  { next_swap_id = 0n
+  ; swaps = (Big_map.empty : (swap_id, swap_info) big_map)
+  }
+
+(* ==== Helpers ==== *)
+
+(* This is just a marker of that the given error should happen
+ * in no circumstances because of invariants of our contract.
+ *)
+[@inline]
+let unexpected_err(err : string) : string = err
+
+let forbid_xtz_transfer : unit =
+  if Tezos.amount = 0tez
+  then ()
+  else failwith "XTZ_TRANSFER"
+
+let fa2_transfer_entrypoint(fa2, on_invalid_fa2 : address * string)
+    : ((FA2.transfer list) contract) =
+  match (Tezos.get_entrypoint_opt "%transfer" fa2 : (FA2.transfer list) contract option) with
+  | None -> (failwith on_invalid_fa2 : (FA2.transfer list) contract)
+  | Some c ->  c
+
+let transfer_asset(from_, to_, on_invalid_fa2 : address * address * string)(asset : fa2_assets)
+    : operation =
+  let transfer_ep = fa2_transfer_entrypoint(asset.fa2_address, on_invalid_fa2) in
+  let token_to_tx(token : fa2_token) : FA2.transfer =
+        { from_ = from_
+        ; txs =
+          [ { to_ = to_
+            ; token_id = token.token_id
+            ; amount = token.amount
+            }
+          ]
+        }
+        in
+  // TODO: maybe expect assets grouped by 'from_' as well? Since FA2 does so.
+  // Probably it may be slightly more efficient
+  // (some checks on 'from_' on FA2's side will be deduplicated),
+  // but would complicate the interface even further.
+  let param = List.map token_to_tx asset.tokens in
+  Tezos.transaction param 0mutez transfer_ep
```

Comment:
I think your performance improvement can be done in-place without needing any change in the contract interface:
```suggestion
let transfer_asset(from_, to_, on_invalid_fa2 : address * address * string)(asset : fa2_assets)
    : operation =
  let transfer_ep = fa2_transfer_entrypoint(asset.fa2_address, on_invalid_fa2) in
  let token_to_tx(token : fa2_token) : FA2.transfer_destination =
        { to_ = to_
        ; token_id = token.token_id
        ; amount = token.amount
        }
      in
  let transfer_txs = List.map token_to_tx asset.tokens in
  let param : FA2.transfer = { from_ = from_; txs = transfer_txs} in
  Tezos.transaction param 0mutez transfer_ep
```
Note: I haven't checked this, but it should work